conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/ServerRequestContext.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.undertow.lib;
+
+import com.google.common.annotations.Beta;
+import com.google.common.collect.ListMultimap;
+import java.util.List;
+import java.util.Optional;
+
+@Beta
+public interface ServerRequestContext {
+
+    /**
+     * Returns all values of the header named {@code headerName}. The name is case insensitive. An empty list is
+     * returned if no such header exists.
+     */
+    List<String> header(String headerName);
+
+    /**
+     * Returns the value of the first header named {@code headerName}. The name is case insensitive.
+     * An {@link Optional#empty()} is returned if no such header exists.
+     */
+    Optional<String> firstHeader(String headerName);
+
+    /**
+     * Returns all query parameters associated with the current request.
+     */
+    ListMultimap<String, String> queryParameters();
```

Comment:
This differs from the header accessors, I expect there to be fewer query parameters, so this should be reasonable, where we headers will be queried more frequently. Unlike headers, there's a use case for iterating through all query parameter names.

Note: this api uses guava ListMultimap. The async feature already uses guava ListenableFuture, and the type is cleaner than `Map<String,List<String>>` so I think it's okay.